EssentialFeed/EssentialFeedTests/Image Comments/API/ImageCommentsMapperTests.swift

```diff
@@ -0,0 +1,81 @@
+//
+// Copyright © 2021 Essential Developer. All rights reserved.
+//
+
+import XCTest
+import EssentialFeed
+
+class ImageCommentsMapperTests: XCTestCase {
+	func test_map_throwsErrorOnNon2xxHTTPResponse() throws {
+		let json = makeItemsJSON([])
+		let samples = [199, 101, 300, 400, 500]
+
+		try samples.forEach { code in
+			XCTAssertThrowsError(
+				try ImageCommentsMapper.map(json, from: HTTPURLResponse(statusCode: code))
+			)
+		}
+	}
+
+	func test_map_throwsErrorOn2xxHTTPResponseWithInvalidJSON() throws {
+		let invalidJSON = Data("invalid json".utf8)
+		let samples = [200, 201, 233, 250, 299]
+
+		try samples.forEach { code in
+			XCTAssertThrowsError(
+				try ImageCommentsMapper.map(invalidJSON, from: HTTPURLResponse(statusCode: code))
+			)
+		}
+	}
+
+	func test_map_deliversNoItemsOn2xxHTTPResponseWithEmptyJSONList() throws {
+		let emptyListJSON = makeItemsJSON([])
+		let samples = [200, 201, 233, 250, 299]
+
+		try samples.forEach { code in
+			let result = try ImageCommentsMapper.map(emptyListJSON, from: HTTPURLResponse(statusCode: code))
+
+			XCTAssertEqual(result, [])
+		}
+	}
+
+	func test_map_deliversItemsOn2xxHTTPResponseWithJSONItems() throws {
+		let item1 = makeItem(
+			id: UUID(),
+			message: "a message",
+			createdAt: ISO8601DateFormatter().date(from: "2020-05-20T11:24:59+0000")!,
+			username: "a username")
+
+		let item2 = makeItem(
+			id: UUID(),
+			message: "another message",
+			createdAt: ISO8601DateFormatter().date(from: "2020-05-20T11:24:59+0000")!,
+			username: "another username")
+
+		let json = makeItemsJSON([item1.json, item2.json])
+		let samples = [200, 201, 233, 250, 299]
+
+		try samples.forEach { code in
+			let result = try ImageCommentsMapper.map(json, from: HTTPURLResponse(statusCode: code))
+
+			XCTAssertEqual(result, [item1.model, item2.model])
+		}
+	}
+
+	// MARK: - Helpers
+
+	private func makeItem(id: UUID, message: String, createdAt: Date, username: String) -> (model: ImageComment, json: [String: Any]) {
+		let item = ImageComment(id: id, message: message, createdAt: createdAt, username: username)
+
+		let json = [
+			"id": id.uuidString,
+			"message": message,
+			"created_at": ISO8601DateFormatter().string(from: createdAt),
+			"author": [
+				"username": username
+			]
+		].compactMapValues { $0 }
```

Comment:
`compactMapValues` is not needed in this context because none of the values are optional. Remove it:

```suggestion
		let json: [String: Any] = [
			"id": id.uuidString,
			"message": message,
			"created_at": ISO8601DateFormatter().string(from: createdAt),
			"author": [
				"username": username
			]
		]
```